src/main/java/works/crossdock/CrossdockServerInboundHandler.java

```diff
@@ -0,0 +1,135 @@
+ * LIABILITY, WHETHER IN AN ACTION OF CONTRACT, TORT OR OTHERWISE, ARISING FROM,
+ * OUT OF OR IN CONNECTION WITH THE SOFTWARE OR THE USE OR OTHER DEALINGS IN
+ * THE SOFTWARE.
+ */
+package works.crossdock;
+
+import com.fasterxml.jackson.core.JsonProcessingException;
+import com.fasterxml.jackson.databind.ObjectMapper;
+import com.google.common.base.Charsets;
+import com.google.common.base.Throwables;
+import io.netty.buffer.Unpooled;
+import io.netty.channel.ChannelFutureListener;
+import io.netty.channel.ChannelHandlerContext;
+import io.netty.channel.SimpleChannelInboundHandler;
+import io.netty.handler.codec.http.DefaultFullHttpResponse;
+import io.netty.handler.codec.http.FullHttpRequest;
+import io.netty.handler.codec.http.FullHttpResponse;
+import io.netty.handler.codec.http.HttpResponseStatus;
+import io.netty.handler.codec.http.HttpVersion;
+import io.netty.handler.codec.http.QueryStringDecoder;
+import java.util.HashMap;
+import java.util.List;
+import java.util.Map;
+import java.util.Map.Entry;
+import java.util.concurrent.CompletableFuture;
+import java.util.concurrent.CompletionStage;
+import works.crossdock.client.Behavior;
+
+public class CrossdockServerInboundHandler extends SimpleChannelInboundHandler<FullHttpRequest> {
+  private final Map<String, Behavior> behaviors = new HashMap<String, Behavior>();
+
+  /** Populates all the different behaviors supported. */
+  public CrossdockServerInboundHandler(Map<String, Behavior> inputBehaviors) {
+    for (Entry<String, Behavior> entry : inputBehaviors.entrySet()) {
+      behaviors.put(entry.getKey(), entry.getValue());
+    }
+  }
+
+  @Override
+  protected void channelRead0(ChannelHandlerContext ctx, FullHttpRequest fullHttpRequest)
+      throws Exception {
+    QueryStringDecoder queryStringDecoder = new QueryStringDecoder(fullHttpRequest.uri());
+    CrossdockRequest request = populateRequest(queryStringDecoder);
+    String behavior = request.getParam("behavior");
+    if (behavior == null) {
+      FullHttpResponse httpResponse =
+          new DefaultFullHttpResponse(HttpVersion.HTTP_1_1, HttpResponseStatus.OK);
+      ctx.writeAndFlush(httpResponse).addListener(ChannelFutureListener.CLOSE);
+      return;
+    }
+    Behavior br = behaviors.get(behavior);
+    runBehavior(br, request)
+        .whenComplete(
+            (response, ex) -> {
+              if (ex != null) {
+                ctx.fireExceptionCaught(ex);
+              } else {
+                writeAndCloseChannel(ctx, response);
+              }
+            });
+  }
+
+  /**
+   * Runs the request against the passed behavior.
+   *
+   * @param br behavior to run against
+   * @param req request to run
+   * @return CompletionStage that contains crossdockResponse
+   * @throws Exception if behavior throws an exception
+   */
+  public CompletionStage<CrossdockResponse> runBehavior(Behavior br, CrossdockRequest req)
+      throws Exception {
+    if (br == null) {
```

Comment:
so Its skipped when the behavior doesn't exist? This is wrong. If the behavior doesn't exist, it should return a hard error. And this function runs a behavior, which by definition cannot be null. The caller for `runBehavior` is responsible for passing a valid behavior object.